transport/src/main/java/com/strategyobject/substrateclient/transport/ws/WsProvider.java

```diff
@@ -167,19 +172,9 @@ public void disconnect() {
         // switch off autoConnect, we are in manual mode now
         this.autoConnectMs = 0;
 
-        try {
-            this.webSocket.updateAndGet(ws -> {
-                if (ws != null) {
-                    ws.close(CloseFrame.NORMAL);
-                }
-
-                return null;
-            });
-
-        } catch (Exception ex) {
-            logger.error("Error disconnecting", ex);
-            this.emit(ProviderInterfaceEmitted.ERROR, ex);
-            throw ex;
+        val ws = this.webSocket;
```

Comment:
There are some moments that confuse me:
1. The method `connect` is asynchronous, but this isn't.
2. The method doesn't consider whether a connection opened.
3. The next call of `connect` won't have expected effect after `disconnect`.

Reply:
About the third point - they should use the same lock or `AtomicReference` for `whenConnected`.

Reply:
Also I would propose to add the test case for the third point - manual reconnect.

Reply:
refactored `disconnect` method as well. also added the test.